proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,242 @@
+pending for distribution.
+
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks for distribution in the next phase.
+
+Currently, on Solana Mainnet Beta with ~550K active stake accounts, it shows
+that epoch reward calculation takes around 10 seconds on average. This will
+make it impossible to perform rewards computation synchronous within one block.
+To make this work, a design of complete async reward computation is required.
+
+However, there are quite a few promising optimizations that can cut down the
+reward computation time. An experiment for reward calculation optimization
+(https://github.com/solana-labs/solana/pull/31193) showed that we can cut the
+reward calculation time plus vote reward distribution time to around 1s. This
+makes synchronous reward computation and asynchronous reward distribution a
+feasible approach. We also believe that there is still more rooms for more
+optimization to further cut down the above timing.
+
+Therefore, the following of the design is based on the above optimization. The
+reward calculation will be performed at the first block of the epoch. Once, the
+full rewards are calculated, the rewards are partitioned into distribution
+chunks stored in the bank, which will then be distributed during the `reward
+distribution` phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+The reward results are sorted by Stake Account address, and randomly shuffled
+with a fast `rng` seeded by current epoch. The shuffled reward results are then
+divided into `M` chunks. This process will ensure that the reward distribution
+is deterministic for the current epoch, while also introducing a degree of
+randomness between epochs.
+
+To minimize the impact on block processing time during the reward distribution
+phase, a total of 4,096 accounts will be distributed per block. The total
+number of blocks needed to distributed rewards is given by the following
+formula to avoid using floating point value arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+### `EpochRewards` Sysvar Account
+
+`EpochRewards` sysvar account records the total rewards and the amount of
+distributed rewards for the current epoch internally. And the account balance
+reflects the amount of pending rewards to distribute.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+
+```
+struct RewardRewards{
+   total_rewards_in_lamport: u64,          // total rewards for the
+                                           // current epoch
+   distributed_rewards_in_lamport: u64,    // distributed rewards for
+                                           // the current epoch
+}
```

Comment:
I think the suggestion that @joncinque and I arrived at yesterday got lost in the comments somehow. We were thinking of adding this new field to the `EpochRewards` sysvar to make it very easy for frontends to determine when the completion of the reward distribution will occur:

```suggestion
    distribution_complete_block_height: u64, // distribution of all staking rewards for the current epoch will be completed before this block height
}
```

Reply:
Yes. sounds good.